vault-tool/src/Network/VaultTool/KeyValueV2.hs

```diff
@@ -0,0 +1,284 @@
+    runVaultRequest,
+    runVaultRequest_,
+    withStatusCodes,
+ )
+import Network.VaultTool.Types (
+    VaultConnection,
+    VaultException (..),
+    VaultMountedPath (..),
+    VaultSearchPath (..),
+    VaultSecretPath (..),
+ )
+
+data VaultSecretMetadata = VaultSecretMetadata
+    { _VaultSecretMetadata_leaseDuration :: Int
+    , _VaultSecretMetadata_leaseId :: Text
+    , _VauleSecretMetadata_renewable :: Bool
+    }
+    deriving (Show, Eq {- TODO Ord -})
+
+instance FromJSON VaultSecretMetadata where
+    parseJSON (Object v) =
+        VaultSecretMetadata
+            <$> v .: "lease_duration"
+            <*> v .: "lease_id"
+            <*> v .: "renewable"
+    parseJSON _ = fail "Not an Object"
+
+data VaultSecretVersion a = VaultSecretVersion
+    { vsvData :: a
+    , vsvMetadata :: VaultSecretVersionMetadata
+    }
+    deriving (Show)
+
+instance FromJSON a => FromJSON (VaultSecretVersion a) where
+    parseJSON = withObject "VaultSecretVersion" $ \v ->
+        VaultSecretVersion
+            <$> v .: "data"
+            <*> v .: "metadata"
+
+data VaultSecretVersionMetadata = VaultSecretVersionMetadata
+    { vsvmCreatedTime :: UTCTime
+    , vsvmDeletionTime :: Maybe UTCTime
+    , vsvmDestroyed :: Bool
+    , vsvmVersion :: Int
+    }
+    deriving (Show)
+
+instance FromJSON VaultSecretVersionMetadata where
+    parseJSON = withObject "VaultSecretVersionMetadata" $ \v ->
+        VaultSecretVersionMetadata
+            <$> v .: "created_time"
+            <*> (v .: "deletion_time" >>= parseOptionalDate)
+            <*> v .: "destroyed"
+            <*> v .: "version"
+      where
+        parseOptionalDate x = parseJSON x <|> pure Nothing
+
+vaultRead ::
+    FromJSON a =>
+    VaultConnection ->
+    VaultSecretPath ->
+    -- | A 'Left' result
+    -- means that the
+    -- secret's "data"
+    -- could not be
+    -- parsed into the
+    -- data structure
+    -- that you
+    -- requested.
+    --
+    -- You will get the
+    -- "data" as a raw
+    -- 'Value' as well as
+    -- the error message
+    -- from the parse
+    -- failure
+    IO (VaultSecretMetadata, Either (Value, String) (VaultSecretVersion a))
+vaultRead conn path = vaultReadVersion conn path Nothing
+
+vaultReadVersion ::
+    FromJSON a =>
+    VaultConnection ->
+    VaultSecretPath ->
+    Maybe Int ->
+    -- | A 'Left' result
+    -- means that the
+    -- secret's "data"
+    -- could not be
+    -- parsed into the
+    -- data structure
+    -- that you
+    -- requested.
+    --
+    -- You will get the
+    -- "data" as a raw
+    -- 'Value' as well as
+    -- the error message
+    -- from the parse
+    -- failure
+    IO (VaultSecretMetadata, Either (Value, String) (VaultSecretVersion a))
+vaultReadVersion conn (VaultSecretPath (mountedPath, searchPath)) version = do
+    let path = vaultActionPath ReadSecretVersion mountedPath searchPath <> queryParams
+    rspObj <-
+        runVaultRequest conn $
+            newGetRequest path
```

Comment:
Ultimately, this is your call, but IMO it's more readable not to split this over multiple lines.  There are a few places where a short line is broken across multiple lines.

Reply:
I'm with you on this one. I think this was my fourmolu plugin auto-formatting things...